### PR TITLE
Add support for embedding of analytics pixels on package pages

### PR DESCRIPTION
**Pull request by alexbiehl**
## Summary

This PR adds support for package maintainers to attach analytics pixels to one or more of their packages in order to better understand the web traffic to their package pages on Hackage. Specifying an analytics pixel URL will result in Hackage automatically embedding the provided URL in an image tag on the main package page. 

## Context

This PR comes from a previous discussion that @aviaviavi, @gbaz, and the hackage-admin email group. The original suggestion for the direction in this PR came @davean, in discussion with the Haskell Foundation regarding approaches for introducing more analytical tools for maintainers within Hackage.

## What not in this PR

Support for analytics pixels to be embedded in generated Haddock documentation is a natural next step and can be included in a future PR.

### CODE REVIEW

File: src/Distribution/Server/Features/AnalyticsPixels/State.hs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    deriving (Show, Eq, Ord, NFData, Typeable, MemSize)
+
+newtype AnalyticsPixelsState = AnalyticsPixelsState
+    {
+        analyticsPixels :: Map PackageName (Set AnalyticsPixel)
+    }
+  deriving (Show, Eq, NFData, Typeable, MemSize)
+
+-- SafeCopy instances
+$(deriveSafeCopy 0 'base ''AnalyticsPixel)
```
Comment (andreasabel) [inline src/Distribution/Server/Features/AnalyticsPixels/State.hs:48]: GHC 9.x's TemplateHaskell seems to process the source top-to-bottom, so the TH invocations need to be put in the right locations.  See the reorderings I performed in 9d4a4900584a87f85f3ca72a8fc907dcfbc1e21b when migrating `hackage-server` to GHC 9.0.

**Review by andreasabel — CHANGES_REQUESTED** [review]
Building with GHC 9.x needs code rearrangements (due to TH).

File: src/Distribution/Server/Features/AnalyticsPixels/State.hs (hunk 1 above)
Comment (alexbiehl) [reply src/Distribution/Server/Features/AnalyticsPixels/State.hs:48]: Thanks for the pointer! Will take care of it

Comment (alexbiehl) [reply src/Distribution/Server/Features/AnalyticsPixels/State.hs:48]: Done!

File: datafiles/templates/Html/analytics-pixels-page.html.st [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+<div id="content">
+
+<h2>Adding a analytics pixel to <a href="/package/$pkgname$">$pkgname$</a></h2>
+
+<p> 
+  Configure an analytics pixel to be automatically loaded on your package’s page on Hackage. 
+  You’ll need an image URL from any external analytics provider, e.g. <a href="about.scarf.sh">Scarf</a>, which is provided 
+  for free and can surface information about web traffic to your package including geographic 
+  distribution, version distribution, and companies.
```
Comment (gbaz) [inline datafiles/templates/Html/analytics-pixels-page.html.st:19]: This text, which explicitly directs users to a specific service and provides promotional copy for it, is not going to work. We can merge without changes here, and I'll just modify it in a subsequent commit, if you prefer. But just wanted to be upfront about that. Ditto for the same text in the other template.

File: src/Distribution/Server/Features/Upload/State.hs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE DeriveDataTypeable, TypeFamilies, TemplateHaskell #-}
+{-# LANGUAGE DeriveDataTypeable, RecordWildCards, TypeFamilies, TemplateHaskell #-}
```
Comment (gbaz) [inline src/Distribution/Server/Features/Upload/State.hs:1]: I think adding this reverse index is a rather heavyweight and unnecessary change. Fetching all packages for a user is already sufficiently efficient. Cf https://hackage.haskell.org/user/BryanOSullivan -- fast!

Comment (alexbiehl) [reply src/Distribution/Server/Features/Upload/State.hs:1]: OK! I will change that then.

Comment (alexbiehl) [reply src/Distribution/Server/Features/Upload/State.hs]: I updated the PR using the GroupIndex mechanics and reverted all the changes to the Upload feature.

File: datafiles/templates/Html/analytics-pixels-page.html.st (hunk 2 above)
Comment (alexbiehl) [reply datafiles/templates/Html/analytics-pixels-page.html.st:19]: I removed the references!

**Comment by alexbiehl** [comment]
Any chance for another round of reviews @gbaz?

**Comment by gbaz** [comment]
It looks good. I  plan to give it a final review and merge before the next redeploy. Just been a bit slammed by other responsibilities!

**Comment by gbaz** [comment]
I merged in preperation for a redeploy, but noticed in testing that the list of pixels already created did not show up properly in the package page. (It does show up in the page that lists all pixels for all packages for a user). So I'm going to comment out links to this functionality for now in the release. A patch fixing that up would be welcome.
